### ISSUE TITLE
Feature request: create a cache in memory like a RAMdisk

### ISSUE THREAD

**Issue by as-muncher** [BiglySoftware/BiglyBT#2126]
* OS and version
* BiglyBT Version Number
* For *nix: Desktop Environment

Java 1.8.0_202 (64 bit)
  Oracle Corporation
c:\program files\biglybt\jre

SWT v4942r22, win32, zoom=100, dpi=120
Windows 10 v10.0, amd64 (64 bit)
B2.7.0.2/4 az3

Hi.  I think uTorrent does some caching in RAM, and I have an encrypted hard drive, not SSD, and the biggest bottleneck has been the hard drive.  Sometimes, I was getting 14 MB/s in download speed, which of course helps a lot!  I turned off "append data to files and reorder pieces as the download progresses" and chose instead "use sparse files when supported by the file system".  That should speed things up, right?  However, it would be lovely if BiglyBT would take some of my RAM, say 1 GB or 2 GB or even more, and read and write from there.  That way, I could get those massive download speeds, and then BiglyBT could write to the disk when the file's done or close to be done and I wouldn't keep getting 100% usage all the time, which then makes BiglyBT lower its downloading speed.  I don't know if that's called "cache thrashing" or what, but it would be helpful if BiglyBT would just keep downloading into the RAM cache and not let the hard drive slow it down.  Then, when the RAM cache is full, then maybe slow down the download speed until the bottlenecks are resolved.

**Comment by parg** [comment]
BiglyBT does have a RAM cache and it helps to reduce disk operations. However, it doesn't scale into the GB range, rather it is designed to work at the piece level.

OS level caching it the way to go - that is aware of all applications attempting to access file content and can act appropriately.

Perhaps you can configure a RAM disk, download to that and then use 'move on complete' to write the data to disk. Of course writing to non-persistent storage has implications around recovery from crashes.

**Comment by as-muncher** [comment]
Hi, thanks for the reply.  OS caching the way to go, eh?  I see so much RAM being unused, and the bottleneck is the hard drive, being encrypted and so it just doesn't write fast enough.  Am I setting the right options to optimize the hard disk usage?  If I'm downloading a 30 GB file, though, I'm not sure how a ramdisk would help, unless I was downloading a file that was less than or equal to the ramdisk size.  But, one can dream big.  :)

**Comment by hglecke** [comment]
Hi:
I explored the same issue with my ancient system. I had to limit transfer speeds in BiglyBT and keep simultaneous downloads to only three. The RAM drive I tried retained the data under most conditions, but limited size remained a problem.
Then I bought a 1TB M.2 drive and a M.2 to PCIE adapter intending to replace the currently used SATA SSD system drive. This promised some limited speed gain. Before using the new M.2 drive as system drive I experimented by letting BiglyBT write exclusively to this drive and move a download when finished to a traditional SATA HD. The result was nothing short of astounding. I could remove all transfer limits from BiglyBT and allow as many as 10 simultaneous downloads.
The speed gain was as much as 10 fold up and down with no disk read/write errors. Finally I was able to use the full internet speed I was paying for.
Needless to say, the M.2 drive remained as the BiglyBT read/write cache. 
Food for thought.
A word of caution!! The N.2 drive type, the N.2 to PCIE adapter and the PCEI slot must be carefully matched for compatibility.
